### ISSUE TITLE
Importing DDI XML from NESSTAR to dataverse for dataset creation

### ISSUE THREAD

**Issue by rmo-cdsp** [IQSS/dataverse#4593]
Hello !

I'm getting some trouble with the api. I am trying to use this command to create datasets via ddi xml files into dataverse:

curl -H "X-Dataverse-key:$API_TOKEN" "$DATAVERSE_URL/api/batch/migrate/?dv=$DATAVERSENAME&path=$PATH_TO_FOLDER"

Of course, every $VARIABLE is replaced with the given values, for my tests. The thing is, I do have errors when the api_token isn't right, or if the dataverse name isn't right. But with the path, no errors is shown, no matter what I write. I actually get this response message : "WORKFLOW_IN_PROGRESS"

Nothing is created on my dataverse. By looking at glassfish logs files, I got this.
validationLog file:
Unexpected Error in handleFile(), file:temp/export.xmlmessage: temp/export.xml, caused by: null at line:

cleanupLog file:
(empty)

A file used for the tests is available here (github doesn't allow xml file, so you will have to change the extension): 
[custom_ddi.txt](https://github.com/IQSS/dataverse/files/1920577/custom_ddi.txt)

If anyone has used this api with success is around here, I would be glad to know what was the good path used :)
And if you have any other intel about this case, it's good to know too !

Thanks for reading, and sorry for my frenchglish ...

EDIT: I discussed about this case with pdurbin, on the dataverse irc channel (http://chat.dataverse.org/). You can see more elements in there: http://irclog.iq.harvard.edu/dataverse/2018-04-17

**Comment by pdurbin** [comment]
@rmo-cdsp nice chatting with you today. Thanks for opening this issue. I linked to it from the long thread we talked about: https://groups.google.com/d/msg/dataverse-community/qsY8swD9Hh0/d84I3SqNAwAJ

First the good news. I was able to import the DDI you attached here after making some tweaks. Here's how the metadata looks in Dataverse:

![dynamiques_de_mobilisation_comprendre_la_formation_des_choix_electoraux_-_vague_1_ 2013 _-_root_-_2018-04-17_19 40 39](https://user-images.githubusercontent.com/21006/38904237-40a2ab96-4277-11e8-8558-50b46c093487.png)

The bad news is that I had to edit the DDI to get it to import. Specifically, I made the changes that appear below:

```
#!/bin/bash
API_TOKEN='f24e501b-459e-474c-8488-48d45f1a8a2e'
DATAVERSE_URL='https://dev1.dataverse.org'
# I think there needs to be a match between the dataverse alias
# you are importing to (i.e. "root") and the folder name.
DATAVERSENAME='root'
PATH_TO_FOLDER="/tmp/$DATAVERSENAME"
mkdir $PATH_TO_FOLDER
cd $PATH_TO_FOLDER
curl -L -s https://github.com/IQSS/dataverse/files/1920577/custom_ddi.txt > /tmp/custom_ddi.orig.xml
cp /tmp/custom_ddi.orig.xml $PATH_TO_FOLDER/edited.xml
echo 'Now edit $PATH_TO_FOLDER/edited.xml...'
echo 'add source="DVN"'
echo 'add type="RELEASED"'
echo 'put "1" in'
#[pdurbin@dvnweb-vm2 root]$ diff custom_ddi.orig.xml custom_ddi.xml 
#56,65c56,57
#<          <verStmt>
#<             <version date="2017-05-04"><![CDATA[Version 2
#< 
#< Des modifications ont été apportées à ce fichier :  
#< 
#< - Renommage de variables
#< - Correction du calcul de variables
#< - Intégration des formules de calcul des variables dans la documentation
#< - Des formats supplémentaires de fichiers de données : SAS, STATA, CSV
#< - Standardisation de la documentation (écriture des filtres etc.)]]></version>
#---
#>          <verStmt source="DVN">
#>             <version date="2017-05-04" type="RELEASED">1</version>
#[pdurbin@dvnweb-vm2 root]$ 
echo 'Exiting early but there's a curl command below.'
exit
curl -H "X-Dataverse-key:$API_TOKEN" "$DATAVERSE_URL/api/batch/migrate/?dv=$DATAVERSENAME&path=$PATH_TO_FOLDER"
```

I'm not sure how clear that is but all of the edits were made to the `verStmt` area. I had to

- add `source="DVN"`
- add `type="RELEASED"`
- change the content of `<version>` to `1` (for version 1)

That at least got it to import, as seen above. There are no files. Also, I'm pretty sure that the folder name needs to be "root" or whatever the alias is of the dataverse you're importing into.

Since there's so much interest in importing NESSTAR data, we should probably endeavor to support this in Dataverse some day. I hope that the tips above at least help unblock the community to the point that we can get feedback on where the bugs are. The main bug seems to be that Dataverse should be able to import the NESSTAR DDI as is without needing to make any changes like I had to above. What other bugs are there. Does what I'm writing here help the community make progress?

**Comment by rmo-cdsp** [comment]
Ah, perfect ! We managed to import a dataset to our dataverse, thanks to you. I admit that an "easier" way of importing in dataverse would be great, but at least we got a way which is great.

Thanks again for your help :+1:

**Comment by pdurbin** [comment]
@rmo-cdsp I'm glad to hear that the comment I made above made enough sense to be helpful. Is there anything else you need? Maybe the easier way that just works without needing to edit the DDI?

**Comment by rmo-cdsp** [comment]
@pdurbin If could not edit XML files, it would be great yes. But an other option would be to be able, through the user interface, to import datas via a DDI XML file (for a regular user side). Also, an entry point in the api, instead of having to put files directly on the dataverse server, would be, for sure, pleasant :)

All of this is for comfort, of course, for now I am happy to have a way to import DDI XML from NESSTAR without having to do all the xml to json process, so I don't need anything more ... for now :D Thanks again.

**Comment by pdurbin** [comment]
@rmo-cdsp ok. Out of curiosity, would you be interested in making a pull request to add any of this functionality? I completely agree that the NESSTAR DDI files shouldn't have to be placed on the file system of the Dataverse server. Using the API, you should be able to just POST the files from anywhere.

**Comment by rmo-cdsp** [comment]
@pdurbin I am, yes. I just have to find some time to do this and think again in Java in my head :grinning:

**Comment by janetm** [comment]
Hi All
I read this issue as I was wondering how using the migration tool dealt with unmapped DDI to Dataverse elements - i can see <version> was commented? 

Anyway the other question is how have you used the migrate API to import the data and other materials from Nesstar into Dataverse 4.x?

Thanks
Janet

**Comment by pdurbin** [comment]
@janetm hi! I'm realizing that your questions above were asked before the Dataverse community meeting. Did you get answers then?

@rmo-cdsp what's your status, please? Still interested in maybe making a pull request some day?

**Comment by rmo-cdsp** [comment]
@pdurbin I intend to work on this feature during next weeks, had no time available before that.

**Comment by pdurbin** [comment]
@rmo-cdsp great! When you're ready to begin, please let us know and we'll move this issue into the new "Community Dev" column at https://waffle.io/IQSS/dataverse

![screen shot 2018-07-17 at 9 52 56 am](https://user-images.githubusercontent.com/21006/42821748-464d16ae-89a7-11e8-8395-6637e68818ea.png)

**Comment by rmo-cdsp** [comment]
@pdurbin I've started to work on the feature. I have some questions about the feature:
- I have found an already existing method of ddi xml file importing feature (postImport in BatchImport.java). It takes a ddi xml file as data posted in the request.  I managed to make it works with Nesstar ddi xml files, by fixing some parts of the code (multiple tags in xml causing error while parsing, dataset versions missing causing errors, and some others). Should I make an api import in the native api only ? (in Dataverses.java for example, instead of batch ?) If so, which precise url should I use ?

- I also have a strange problem in my dev environnement (4.9.2) : I can't publish datasets on my own dataverse dev installation. I kept the basic installation settings and I thought there was no problems by using the ezid apitest defined there. Something has changed ? (It worked with my previous 4.9.1 dev version, if I'm not mistaking) I had to comment some part of the code to prevent doi testing before publishing datasets, to do my tests :S

**Comment by pameyer** [comment]
@rmo-cdsp I'm not sure exactly when this changed, but relatively recently I noticed being unable to publish datasets in my dev environment using the EZID apitest credentials unless the site url configuration was set (which wasn't previously the case).

**Comment by pdurbin** [comment]
@rmo-cdsp thanks for working on this! I just dragged this issue to the new "Community Dev" column at https://waffle.io/IQSS/dataverse and assigned it to you.

With regard to your questions:

- You should do something like `$SERVER_URL/api/dataverses/$DV_ALIAS/datasets/:import?pid=$PERSISTENT_IDENTIFIER&release=yes` as documented at http://guides.dataverse.org/en/4.9.2/api/native-api.html#import-a-dataset-into-a-dataverse which is what @pameyer was talking about. That is to say, yes, it should go into the Native API and it should use a format more or less like the one above. Here's the code: https://github.com/IQSS/dataverse/blob/v4.9.2/src/main/java/edu/harvard/iq/dataverse/api/Dataverses.java#L240
- Can you please open a new issue about being unable to publish datasets? Thanks!

**Comment by rmo-cdsp** [comment]
@pdurbin I am back to the subject and wanted to fix the EZID problem with @pameyer solution (I will open a new issue if needed), but I ran into a new problem (when I merged my code with the lastest develop branch commits), that your are mentioning on irc, about a login problem. So I may wait a little bit before coming back about this ticket :S

**Comment by pdurbin** [comment]
@rmo-cdsp thanks for the quick chat at http://irclog.iq.harvard.edu/dataverse/2018-09-25#i_73255 . Please keep us posted on if you're having trouble with your dev environment. We definitely don't want you to be blocked!

**Comment by pdurbin** [comment]
@rmo-cdsp hi! I'm just checking in to see if you need any help with your dev environment or if you've pushed your branch yet. Please note that we've released Dataverse 4.9.3 so merging the latest from "develop" into your branch is definitely recommended.

**Comment by rmo-cdsp** [comment]
@pdurbin Just tested my thing with 123 ddi files from our data servers. 101 could be properly imported, 22 still has problems. I have 3 problems:
- Countries string not part of the controlled vocabulary (France is ok, but not "FRANCE", not "European Union", etc). Gonna check if it is a problem on our datas as ddi or else. I'll try to do an error message for this side.
- Multiple depositors : it seems ddi allows to have multiple depositors. But not Dataverse. I don't know yet what to do: fail the import and answer back that you can't have multiple depositors for Dataverse ? Or should the multiplicity of depositor as a Dataverse field must be redefined ? Maybe the first case is the easiest one for now to do ?
- An unknow error about an XML parse error ... but I'll find out why soon

So, I'm mostly dealing with "datas" problems. I want to finish this part before merging with 4.9.4 and share my code for revision/thoughts :)

**Comment by kcondon** [comment]
@rmo-cdsp I'm about to test this. I'm assuming you are using Nesstar 4.0 and exporting your datasets in DDI format? Do you have some samples I can test with? Also, do you have any updates on the three issues mentioned in your previous comment?

**Comment by rmo-cdsp** [comment]
@kcondon About my 3 points:
- For now, countries has to be in english and respect your controlled vocabularies. If it's snot the case, an error message will say so. I couldn't move forward on this, but maybe we should use country codes in order to handle the different languages people would use in their xml. On our side, it is something we have to work on (country name controlled vocabulary).
- For multiple depositors, an issue as been created ( #5164 ) and a WIP pull request has been made (#5223 ). For now, depositor is a unique field for dataverse, so ddi files with multiple depositors won't work and a message will say so. In my pull request (#5164 ), I has to check it multiple depositors don't mess with code (@pdurbin made a list of possibilities in the issue). To note, some other ddi fields has to be multiple, so it's not only related to depositors. Ssome of thoses fields are indicated in the issue.
- The unknown error has been fixed and here is why: https://github.com/IQSS/dataverse/pull/5170/files#diff-c494f7f5bb9a2cee025cb9e4217ec25eR133

An other point, I still has no answer for my comment here (nothing critical):
https://github.com/IQSS/dataverse/pull/5170/files#r224506867

If you want ddi xml files samples, you can check our public metadatas repo here: http://nesstar.sciences-po.fr/webview/

By navigating , you can found datasets and will be able to download files (as pdf or xml, choose xml) by clicking on the floppy disk icon.

**Comment by kcondon** [comment]
@rmo-cdsp Thanks! That's helpful. Will finish up tonight and merge or post any questions. Regards.

**Comment by rmo-cdsp** [comment]
@kcondon I will see that tomorrow to fix it. I wonder if the "/datasets/:import" import point has the same problem (given that the process is the same).

**Comment by kcondon** [comment]
@rmo-cdsp I had tested it when it was made and it worked but required a valid PID. I can retest it to be sure.

**Comment by rmo-cdsp** [comment]
@kcondon I had to edit the xml files with other PID for testing, because none had one compatible with my dataverse installation :S

**Comment by kcondon** [comment]
@rmo-cdsp Would you fix the super user issue? I'm looking again at the PID/release= yes problem but I believe it is out of scope for this. Thanks! Kevin

**Comment by rmo-cdsp** [comment]
@kcondon I can do that. It may be hard given my planning tomorrow (holiday next week) but I'll try to do that fast ;)

**Comment by rmo-cdsp** [comment]
@kcondon superuser test added and pushed

 I am taking a look at the PID release:
- if I indicate a doi that already exist (other dataset with the doi in the instance), it will crash (meaning returning bad html in the request) and say in the server log that doi is already used
- if I indicate a doi that doesn't already exist (not in other dataset with the doi in the instance), I got a message to the request stating "Dataset with identifier 'FK2/AAADAA', protocol 'doi' and authority '10.5072' already exists" (my id was doi:10.5072/FK2/4D7DAA)

I don't think I will be able to fix this today, so it may be in a week :S

**Comment by kcondon** [comment]
@rmo-cdsp No worries, take what time you need. I realized after testing and thinking more about it, the options are more involved:
json:
pid provided on command line, or in file but required, release=yes optional. pid has to preexist outside of current system and be accessible/ resolvable and be of a type know to us in order to verify.
ddi:
pid provided on command line or in file or not provided, release=yes optional. pid if provided has to preexist outside of current system and be accessible/ resolvable and be of a type known to us in order to verify.

When passing it a PID with one of the methods, it must actually use that PID when creating the dataset (obviously) but I need to confirm that, even if release=no.

For testing, we use a valid, published doi from  our production namespace on our test box. Since the update credentials are different, it should not interfere in any way.

**Comment by kcondon** [comment]
@rmo_cdsp I've done some more testing:

-confirmed the super-user check is working, thanks
-found the ddi import fails if you specify a PID on the command line
-was not sure how to specify the PID in ddi but I am assuming it fails too
-works if do no specify PID and release=no or not specified.
-If no PID and release=yes, throws 500 error about createtime being null.

compared to json import:
-PID always required
-Specifying the PID works on the command line but not in file
-release=yes/no works

I believe the use cases for ddi import and json import might be different due to who created them. Json import assumes valid PID from another system being imported into Dataverse and likely released immediately. If for some reason the imported PID namespace is different from Dataverse, then must release immediately or cannot publish and cannot publish new versions due to not having credentials to different imported namespace.

Please correct me if I'm wrong but I believe your ddi use case is to import ddi exports from Nesstar where you do not have a doi  or PID that you are preserving and may want to review the import before publishing. That is the use case that is currently working correctly.

My question is one of use case and technical design. What should both endpoints be doing and should they behave the same? It might help to have a developer from the Dataverse team comment here.

@tcoupin at https://github.com/IQSS/dataverse/issues/5104 was also working on some aspect of json import so I am copying him here. You might want to check in with them.

**Comment by tcoupin** [comment]
> I believe the use cases for ddi import and json import might be different due to who created them. Json import assumes valid PID from another system being imported into Dataverse and likely released immediately. If for some reason the imported PID namespace is different from Dataverse, then must release immediately or cannot publish and cannot publish new versions due to not having credentials to different imported namespace.

You should take a look at: https://github.com/IQSS/dataverse/issues/5104#issuecomment-433124385

What you said is true for now but I think it is not the expected behavior. As @djbrooke said, the imported PID should have the same namespace as Dataverse.

**Comment by pdurbin** [comment]
@rmo-cdsp @tcoupin as I was saying to @kcondon yesterday we might want to take a step back and talk about the user cases for import.

Because @pameyer is on our floor twice a week we're quick familiar with his import needs. He has a system with existing DOIs at https://data.sbgrid.org that he's migrating to Dataverse, keeping the same DOIs and the same namespace and has both systems configured with the same credentials for the DOI provider. As I understand it, when he migrates datasets, they will be migrated in a published state (release=yes). These docs apply to his use case: http://guides.dataverse.org/en/4.9.4/api/native-api.html#import-a-dataset-into-a-dataverse

@donsizemore has been making a little noise about his import use case for IPUMS XML at http://irclog.iq.harvard.edu/dataverse/2018-10-24#i_76954

These are the questions that are on my mind for the import use cases:

- Will Dataverse be minting a new DOI?
- Is there an old DOI that needs to be preserved? If so, should it become the main DOI, which would require that it's the same DOI provider namespace and credentials? Or should an old DOI be preserved as Other ID in the metadata or in the new "alternativepersistentidentifier" database table?
- Will the dataset be imported in a published or unpublished state?
- What format needs to be imported? DDI? JSON? What flavor of DDI or JSON?

**Comment by pdurbin** [comment]
Just a heads up the import worked fine for @flaviofrancisco in #5258

**Comment by kcondon** [comment]
Do you know what exactly he did?

**Comment by pdurbin** [comment]
@kcondon nope but I just asked at https://github.com/IQSS/dataverse/issues/5258#issuecomment-435090570

**Comment by rmo-cdsp** [comment]
Back to business !
So, @kcondon , about our use of ddi import:
for now, our NESSTAR DDI datas don't have DOI, so we don't have to give a specific DOI, we will take the one generated by our dataverse DOI provider (which is our organization provider anyway). We have some intern identifiers, which are considered as "other identifier" by dataverse, which is fine.
But we also have DDI metadatas that are not from NESSTAR, and some of it do have a DDI already provided by the same DOI provider used by our dataverse instance.

So, this is why I think multiple options (4 at least, if I count well) for DDI import is necessary. You may have metadatas ready to publish in your DDI, with or wthout DOI, or you may have to change some things before, with or without DOI. Also, in my case mostly, my data managers will not be responsible of the DDI import (command line function is not part of their workflow) but will be responsible of the publication, manually looking at the fields indicated in the datas, and edition if that's the case.
Here is the usecase for us. For external DOI, I don't know what is your "policy" in dataverse app.

 I am looking at the problems we got when we specify a DOI in the ddi import (createtime at null and datacite fail check) and will come back with what I could found about it.

**Comment by rmo-cdsp** [comment]
@kcondon Problems should be fixed now. I can properly : 
- create a dataset with doi and as draft
- create a dataset with doi and as released
- create a dataset without doi and as draft
- create a dataset without a doi and as released

I hope this time is the good one ! :D

**Comment by kcondon** [comment]
@rmo-cdsp Thanks! Works great!
